Creating custom policy controller [WIP]

dockerymick opened the pull request:
https://github.com/open-cluster-management/backlog/issues/1224

dockerymick committed:
Creating custom policy controller

dockerymick committed:
made a few more changes

ycao56 commented on security/create_policy_ctrl.adoc:
```diff
@@ -0,0 +1,248 @@
+Your policy controller is deployed.
+
+[#deleting-a-policy-controller]
+=== Deleting a policy controller
```
the command in this section deletes policy not policy controller

dockerymick replied:
@willkutler may you provide the correct command?

dockerymick replied:
@ycao56 Will shared that he wasn't sure which command should be used to delete a policy controller. May you provide the info?

ycao56 commented on security/policy_controllers.adoc:
```diff
@@ -5,8 +5,9 @@ Use the Red Hat Advanced Cluster Management for Kubernetes governance and risk p
 
-Create and customize policies with the out of the box policy templates.
-*Important*: Some policy controllers might not support the enforce feature.
+Create and customize poolicy controllers and policies with the out of the box policy templates. Policy controllers monitor and report whether your cluster is compliant with policy. It can also enforce the policy to bring the cluster status to compliant. See xref:../create_policy_ctrl.adoc#creating-a-custom-policy-controller[Creating a custom policy controller] for more information.
```
shouldn't mix custom policy controllers with out of box policies. this should be a separate section. e.g.
Bring your own policies by creating custom policy and policy controller

ycao56 commented on security/create_policy_ctrl.adoc:
```diff
@@ -0,0 +1,248 @@
+[#managing-custom-policy-controllers]
+= Managing custom policy controllers
```
We probably need to reorganize the sections. We should explain
1. when and why to write a custom policy and policy controller
2. how to do it
3. how to deploy it and integrate with grc dashboard

dockerymick replied:
@ycao thanks for your suggestions. We have not specified when and why the policy controller should be created, however I suggest that the explanation is added to the security/policy_controllers.adoc file

willkutler replied:
here's how you'd use the new policy controller with the grc dashboard:
1. deploy the controller
2. go to the grc dashboard and click "create policy"
3. edit the yaml so that an instance of your custom policy is listed under `policy-templates`
4. now any violations generated by that custom policy will show up as violations on the parent policy

dockerymick committed:
Update policy_controllers.adoc

Making some updates

dockerymick committed:
Update create_policy_ctrl.adoc

some reorganization

dockerymick commented:
Still need to update the PR. Per @willkutler suggest, we should add a section named `Integrating custom policy controller with the Governance and risk dashboard` and add steps

dockerymick committed:
Update create_policy_ctrl.adoc

I added the section `Integrating the policy controller...` however, I suggest that we add step 7 and reference the user to the correct page

ycao56 commented:
@dockerymick We need to talk about the purpose of this doc. I am still not comfortable with the content. We already have a blog talking about how to write a custom policy controller. Why do we want to duplicate the information? Also the content of the blog was not 100% percent port over to this doc. I am confused about the intention.

dockerymick commented:
@ycao56 The purpose of this doc is to document only task for the policy controller. The requirement for RH content team is to utilize modular documentation, where concepts and task are separate. Concepts for the policy controller are on the Policy controller page and the task are on Managing custom policy controllers. Even though there is a blog that explains creating a custom policy controller, it was requested that the info also be added into the customer portal documentation for 2.0.

dockerymick committed:
Update create_policy_ctrl.adoc

review with yu can will, still a work in progress

dockerymick committed:
Update create_policy_ctrl.adoc

Made some updates, still not sure if delete policy controller should be removed.

dockerymick committed:
Update create_policy_ctrl.adoc

ycao56 commented on security/create_policy_ctrl.adoc:
```diff
@@ -0,0 +1,202 @@
+    listKind: SamplePolicyList
+    plural: samplepolicies
+    singular: samplepolicy
+----
```
there are more steps required to update policy see
<img width="731" alt="image" src="https://user-images.githubusercontent.com/16092291/87972095-58ae2600-ca94-11ea-90d2-740cbe6560d2.png">

ycao56 commented on security/create_policy_ctrl.adoc:
```diff
@@ -0,0 +1,202 @@
+
++
+----
+kubectl describe SamplePolicy example-samplepolicy -n default
```
you need to run policy controller and install crd and cr before you can check the result. see
<img width="681" alt="image" src="https://user-images.githubusercontent.com/16092291/87972291-a3c83900-ca94-11ea-976d-ac4fbe762a21.png">

ycao56 commented on security/create_policy_ctrl.adoc:
```diff
@@ -0,0 +1,202 @@
+
+. Update the policy rules and policy logic to introduce new rules for your policy controller. Complete the following steps:
+
+.. Add new fields in your YAML file by updating the `SamplePolicySpec`. Your specification might resemble the following content:
```
steps are missing to update policy controller

<img width="733" alt="image" src="https://user-images.githubusercontent.com/16092291/87972386-c8241580-ca94-11ea-9fb8-7aae566b44bd.png">

ycao56 commented on security/create_policy_ctrl.adoc:
```diff
@@ -0,0 +1,202 @@
+kubectl get pod -n <namespace>
+----
+
+. You must integrate your policy controller by creating a `policy-template` for the controller to monitor. For more information, see xref:../security/create_policy.adoc#creating-a-cluster-security-policy-from-the-console[Creating a cluster security policy from the console]. 
```
do we have instructions on how user can use its own policy template and create policy through UI?

dockerymick replied:
@ycao We have instructions on how to create a policy from the console here: https://github.com/open-cluster-management/rhacm-docs/blob/2.0_stage/security/create_policy.adoc

dockerymick committed:
Update create_policy_ctrl.adoc

made updates from yu's suggestions and added steps to scale the deployment

dockerymick committed:
Update create_policy_ctrl.adoc

dockerymick committed:
Update create_policy_ctrl.adoc

dockerymick committed:
Update create_policy_ctrl.adoc
